listenbrainz/db/feedback.py

```diff
@@ -155,3 +156,35 @@ def get_feedback_count_for_recording(recording_msid: str):
         count = int(result.fetchone()["value"])
 
     return count
+
+
+def get_feedback_for_multiple_recordings_for_user(user_id: int, recording_list: List[str]) -> List[Feedback]:
+    """ Get a list of recording feedback given by the user for given recordings
+
+        Args:
+            user_id: the row ID of the user in the DB
+            recording_list: list of recording_msid for which feedback records are to be obtained
+                            - if record is present then return it
+                            - if record is not present then return a pseudo record with score = 0
+
+        Returns:
+            A list of Feedback objects
+    """
+
+    args = {"user_id": user_id, "recording_list": recording_list}
+    query = """ WITH rf AS (
+                  SELECT user_id, recording_msid::text, score
+                    FROM recording_feedback
+                   WHERE recording_feedback.user_id=:user_id
+                )
+                SELECT COALESCE(rf.user_id, :user_id) AS user_id, "user".musicbrainz_id AS user_name,
+                       rec_msid AS recording_msid, COALESCE(rf.score, 0) AS score
```

Comment:
Okay, so I have been reading the feedback (db + API + react) code for some time now and there are a few things that must be changed. You can either work on this
or open a ticket so that these *can be buggy in future features* can catch someone's attention.

(I wasn't able to find all the PRs related to feedback work so I am commenting everything here)

1. There should be separate endpoints for feedback submit and feedback delete. The following json will ideally be interpreted as a submission irrespective of the score. {'recording_msid': 'xxx', score: '0'}
Keeping 1, -1 for submission and 0 for deletion and passing through the same endpoint is not a good idea. Also, you are simply checking in the feedback_api script if the submitted score is in [-1, 0, 1] but there is no constraint in the DB to ensure the same. I would recommend making separate endpoint for feedback submission and deletion and also to have DB constraints for the score submitted.

2. You have allowed -1, 0, 1 as valid score values to be submitted to the API endpoint. But here in this query, you are returning a pseudo score as 0 for recording msids that the user has not rated. Then in react at some places, you have used 0 to delete feedback via the endpoint and at some places, you have used 0 to reflect that the feedback for that recording was not submitted. The whole usage of 0 is very confusing. Even though the code works perfectly rn, we may have bugs in the future which are because of this confusing use of 0 or are difficult to track because of this confusing use of 0.

3. Almost at all the places the previous state has been mutated like this https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/static/js/src/RecentListens.tsx#L520. You should ideally make a copy of the previous state and then do setState.